bluetooth: host: Add support for processing CS subevent results

olivier-le-sage opened the pull request:
Adds support for:
- LE CS Subevent Result event
- LE CS Test End Complete event

For now, recombination of subevent results with more steps than could fit within a single HCI event is not supported, and such events are discarded.

bugo-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -413,6 +413,109 @@ struct bt_conn_le_cs_fae_table {
+		 *  indicates that the role is not the initiator, or that the
+		 *  frequency compensation value is unavailable.
+		 */
+		uint16_t frequency_compensation;
```
Should we report this in a more user friendly format ie. int16_t? Or we want to keep it raw in case the user algorithm uses it in the raw format?

olivier-le-sage replied:
Thought process: I saw the RAS spec wants it like this 
![format freq compensation](https://github.com/user-attachments/assets/3f50f727-c8f3-4e49-b5d8-e491648d6873)

bugo-nordic commented on include/zephyr/bluetooth/cs.h:
```diff
@@ -569,6 +603,35 @@ int bt_cs_read_remote_fae_table(struct bt_conn *conn);
+ *
+ * @param step_data_buf Pointer to a buffer containing the step data.
+ * @param func Callback function which will be called for each step data found.
+ *             The callback sould return true to continue parsing, or false to stop.
```
```suggestion
 *             The callback should return true to continue parsing, or false to stop.
```

bugo-nordic commented on subsys/bluetooth/host/cs.c:
```diff
@@ -326,4 +334,134 @@ int bt_cs_start_test(const struct bt_cs_test_param *params)
+
+	evt = net_buf_pull_mem(buf, sizeof(*evt));
+
+	if (evt->subevent_done_status == BT_HCI_LE_CS_SUBEVENT_DONE_STATUS_PARTIAL) {
```
Not too important at this stage but I think this should be moved to below line 367

bugo-nordic commented on subsys/bluetooth/host/cs.c:
```diff
@@ -326,4 +334,134 @@ int bt_cs_start_test(const struct bt_cs_test_param *params)
+}
+
+void bt_cs_step_data_parse(struct net_buf_simple *step_data_buf,
+			bool (*func)(struct bt_cs_subevent_step *step))
```
Would be good to fix the alignment in here and other newly added lines

bugo-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1677,6 +1783,17 @@ struct bt_conn_cb {
+	 * @param conn Connection objects.
+	 * @param result Subevent results
+	 */
+	void (*le_subevent_data_available)(struct bt_conn *conn,
```
```suggestion
	void (*le_cs_subevent_data_available)(struct bt_conn *conn,
```

sean-madigan commented on subsys/bluetooth/host/cs.c:
```diff
@@ -344,6 +351,74 @@ int bt_le_cs_start_test(const struct bt_le_cs_test_param *params)
+	}
+
+	if (evt->subevent_done_status == BT_HCI_LE_CS_SUBEVENT_DONE_STATUS_PARTIAL) {
+		LOG_WRN("Discarded incomplete CS subevent results.");
```
Do we need a `bt_conn_unref(conn);` here? I guess this is basically todo anyway so maybe not necessary

olivier-le-sage replied:
We might as well. Originally I think I solved this by making this check before the conn handle lookup. So I'm thinking I'll just move it back up

bugo-nordic replied:
I'd still prefer leaving this here and adding `bt_conn_unref`. The reason is that the logic will have a branch here to handle partial results. It doesn't make sense to make handle that branch without making the important checks above first.

olivier-le-sage replied:
I was guessing we would end up removing this branch completely and reworking this function later. For now I think we can just keep it simple without if (conn != null) {unref()} and revisit later

olivier-le-sage commented on subsys/bluetooth/host/cs.c:
```diff
@@ -241,6 +242,12 @@ void bt_hci_le_cs_read_remote_fae_table_complete(struct net_buf *buf)
 	bt_conn_unref(conn);
 }
 
+int bt_le_cs_test_cb_register(struct bt_le_cs_test_cb cb)
```
what do you think of a BT_CHANNEL_SOUNDING_TEST ?

jhedberg replied:
> what do you think of a BT_CHANNEL_SOUNDING_TEST ?

Would it have `depends on BT_TESTING`? At least until now I believe we've grouped all testing related Kconfig options behind `BT_TESTING`. The bluetooth tester app (used for qualification) enables `BT_TESTING` so it shouldn't be an issue there. The Bluetooth shell also enables it for some use cases (see e.g. `tests/bluetooth/shell/audio.conf`)

olivier-le-sage replied:
I was thinking of doing the same as with BT_ISO_TEST_PARAMS (no strong dependencies). No strong opinion however, I'm not so familiar with the use-cases for BT_TESTING

jhedberg replied:
`BT_ISO_TEST_PARAMS` doesn't seem to be guarding for any public APIs (functions), so I don't think it's quite the same. *If* we will tie these CS features to `BT_TESTING` I don't think we necessarily need a separate option, since it doesn't really provide added value. @jori-nordic @alwa-nordic any thoughts here?

olivier-le-sage replied:
the nice part of BT_CHANNEL_SOUNDING_TEST is I think it makes it easier to add a corresponding controller kconfig, which can then have something like: `default n if !BT_CHANNEL_SOUNDING_TEST` (which can be used to remove controller code implementing these commands)

by default the test commands need to be included in the controller as they're mandatory by spec

jhedberg replied:
Sure, as long as you don't forget that the controller doesn't have to be part of the same build as the host, e.g. if there's a physical transport between them and they're on different cores.

olivier-le-sage replied:
Yes, that's true.

olivier-le-sage replied:
@jhedberg In any case I've pushed what I imagined this could look like, does it seem ok?

jhedberg replied:
> @jhedberg In any case I've pushed what I imagined this could look like, does it seem ok?

I was thinking that this might still `depends on BT_TESTING` as well. @alwa-nordic any thoughts on this?

alwa-nordic replied:
I have some confidence this does not belong in BT_TESTING. My understanding is we use BT_TESTING to do white-box testing in CI. It exposes some interfaces that allow observing or altering something about internals of the stack. These interfaces must be considered implementation-leaking and very unstable.

`config BT_TESTING` says "This option enables *custom* Bluetooth testing interface." (emphasis mine).

If this is an HCI function, then it's not custom. And if this is useful for testing interoperability and RF environment, then it may be useful for end-users. And I don't want end-users to rely on BT_TESTING.

alwa-nordic replied:
Said plain: I would like BT_TESTING to be owned by, and only available to, the Host (or at least only in-tree in Zephyr).

jhedberg commented on subsys/bluetooth/Kconfig:
```diff
@@ -196,6 +196,13 @@ config BT_CHANNEL_SOUNDING
+config BT_CHANNEL_SOUNDING_TEST
+	bool "Channel Sounding Test support"
+	select EXPERIMENTAL
+	select BT_CHANNEL_SOUNDING
```
Shouldn't this be `depends on BT_CHANNEL_SOUNDING`?

jhedberg commented on subsys/bluetooth/Kconfig:
```diff
@@ -196,6 +196,13 @@ config BT_CHANNEL_SOUNDING
 	  Enable support for Bluetooth 6.0 Channel Sounding feature.
 
+config BT_CHANNEL_SOUNDING_TEST
+	bool "Channel Sounding Test support"
```
`select EXPERIMENTAL` should be accompanied by " [EXPERIMENTAL]" at the end of the option name.

olivier-le-sage committed:
bluetooth: host: Update text for BT_CHANNEL_SOUNDING kconfig

Include EXPERIMENTAL tag in the kconfig text and remove the word
"support" to align with other options

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>

olivier-le-sage committed:
bluetooth: host: Add CONFIG_BT_CHANNEL_SOUNDING_TEST

Makes it easier to make sure the CS test code is removed if desired.
It should hopefully allow for a clean split in general.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>

olivier-le-sage committed:
bluetooth: host: Add support for processing CS subevent results

Adds support for:
- LE CS Subevent Result event
- LE CS Test End Complete event

For now, recombination of subevent results with more steps than could
fit within a single HCI event is not supported, and such events
are discarded.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>
